File: frontend/src/types/screen.ts
// ===== 喵喵選股 v1 API Types =====
// 新架構專用型別，與現有 types/index.ts 共存

// 篩選規則
export interface Rule {
    type: 'indicator' | 'fundamental' | 'chip';
    field: string;
    operator: '>' | '<' | '=' | '>=' | '<=' | 'CROSS_UP' | 'CROSS_DOWN';
    target_type: 'value' | 'field';
    target_value: number | string;
}

// 自訂公式
export interface Formula {
    name: string;
    formula: string;
}

// 篩選請求
export interface ScreenRequest {
    logic: 'AND' | 'OR';
    rules: Rule[];
    custom_formulas: Formula[];
}

// 篩選結果
export interface TickerResult {
    ticker_id: string;
    name: string;
    market_type?: string;
    industry?: string;
    close?: number;
    change_percent?: number;
    volume?: number;
    ma5?: number;
    ma10?: number;
    ma20?: number;
    ma60?: number;
    rsi14?: number;
    pe_ratio?: number;
    eps?: number;
    foreign_buy?: number;
    trust_buy?: number;
    margin_balance?: number;
    // 延伸指標
    turnover?: number;
    avg_volume_20?: number;
    avg_turnover_20?: number;
    lower_shadow?: number;
    lowest_lower_shadow_20?: number;
    wma10?: number;
    wma20?: number;
    wma60?: number;
    market_ok?: boolean;
}

export interface ScreenResponse {
    matched_count: number;
    data: TickerResult[];
    logic: string;
}

// K 線
export interface KlineCandle {
    date: string;
    open?: number;
    high?: number;
    low?: number;
    close?: number;
    volume?: number;
}

export interface KlineResponse {
    ticker_id: string;
    name: string;
    period: string;
    candles: KlineCandle[];
    indicators: Record<string, (number | null)[]>;
}

// 策略
export interface Strategy {
    id: number;
    name: string;
    rules_json: Record<string, any>;
    alert_enabled: boolean;
    created_at?: string;
    updated_at?: string;
}

export interface StrategyCreate {
    name: string;
    rules_json: Record<string, any>;
    alert_enabled?: boolean;
    line_notify_token?: string;
}

export interface StrategyUpdate {
    name?: string;
    rules_json?: Record<string, any>;
    alert_enabled?: boolean;
    line_notify_token?: string;
}

// 股票資訊
export interface TickerInfo {
    ticker_id: string;
    name: string;
    market_type?: string;
    industry?: string;
}

// 篩選欄位選項
export type FieldCategory = 'indicator' | 'fundamental' | 'chip';

export interface FieldOption {
    label: string;
    value: string;
    category: FieldCategory;
}

// 預設可用欄位
export const AVAILABLE_FIELDS: FieldOption[] = [
    // 技術面
    { label: '收盤價', value: 'close', category: 'indicator' },
    { label: '開盤價', value: 'open', category: 'indicator' },
    { label: '最高價', value: 'high', category: 'indicator' },
    { label: '最低價', value: 'low', category: 'indicator' },
    { label: '成交量', value: 'volume', category: 'indicator' },
    { label: 'MA5', value: 'ma5', category: 'indicator' },
    { label: 'MA10', value: 'ma10', category: 'indicator' },
    { label: 'MA20', value: 'ma20', category: 'indicator' },
    { label: 'MA60', value: 'ma60', category: 'indicator' },
    { label: 'RSI(14)', value: 'rsi14', category: 'indicator' },
    { label: '漲跌幅%', value: 'change_percent', category: 'indicator' },
    // 延伸指標
    { label: '成交值', value: 'turnover', category: 'indicator' },
    { label: '20日均量', value: 'avg_volume_20', category: 'indicator' },
    { label: '20日均成交值', value: 'avg_turnover_20', category: 'indicator' },
    { label: '下引價', value: 'lower_shadow', category: 'indicator' },
    { label: '近20日下引價最低(前日基準)', value: 'lowest_lower_shadow_20', category: 'indicator' },
    { label: '週MA10', value: 'wma10', category: 'indicator' },
    { label: '週MA20', value: 'wma20', category: 'indicator' },
    { label: '週MA60', value: 'wma60', category: 'indicator' },
    { label: '大盤多頭OK', value: 'market_ok', category: 'indicator' },
    // 基本面
    { label: '本益比', value: 'pe_ratio', category: 'fundamental' },
    { label: 'EPS', value: 'eps', category: 'fundamental' },
    // 籌碼面
    { label: '外資買賣超', value: 'foreign_buy', category: 'chip' },
    { label: '投信買賣超', value: 'trust_buy', category: 'chip' },
    { label: '融資餘額', value: 'margin_balance', category: 'chip' },
];

// 運算子選項
export interface OperatorOption {
    label: string;
    value: string;
}

export const AVAILABLE_OPERATORS: OperatorOption[] = [
    { label: '大於 (>)', value: '>' },
    { label: '小於 (<)', value: '<' },
    { label: '等於 (=)', value: '=' },
    { label: '大於等於 (>=)', value: '>=' },
    { label: '小於等於 (<=)', value: '<=' },
    { label: '黃金交叉 ↑', value: 'CROSS_UP' },
    { label: '死亡交叉 ↓', value: 'CROSS_DOWN' },
];

File: backend/app/schemas/screen.py
"""
Screen Schemas — 篩選 API 請求/回應模型
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, Union, Literal
from datetime import date


class Formula(BaseModel):
    """自訂公式"""
    name: str = Field(..., description="公式名稱", examples=["avg_ma"])
    formula: str = Field(..., description="公式表達式", examples=["(ma5 + ma10 + ma20) / 3"])


class Rule(BaseModel):
    """篩選條件"""
    type: Literal["indicator", "fundamental", "chip"] = Field(
        ..., description="條件類型"
    )
    field: str = Field(..., description="欄位名稱", examples=["close", "pe_ratio", "foreign_buy"])
    operator: Literal[">", "<", "=", ">=", "<=", "CROSS_UP", "CROSS_DOWN"] = Field(
        ..., description="運算子"
    )
    target_type: Literal["value", "field"] = Field(
        "value", description="目標類型：數值或另一個欄位"
    )
    target_value: Union[float, str] = Field(
        ..., description="目標值 (數值) 或目標欄位名稱 (字串)"
    )


class ScreenRequest(BaseModel):
    """篩選請求"""
    logic: Literal["AND", "OR"] = "AND"
    rules: list[Rule] = Field(default_factory=list)
    custom_formulas: list[Formula] = Field(default_factory=list)

    # config.py 已定義 max_rules_per_request / max_formulas_per_request，
    # 但過去未在 API 邊界強制執行 — 海量規則/公式可造成 CPU DoS
    @field_validator("rules")
    @classmethod
    def _cap_rules(cls, v):
        from config import get_settings
        cap = get_settings().max_rules_per_request
        if len(v) > cap:
            raise ValueError(f"規則數量超過上限 ({cap})")
        return v

    @field_validator("custom_formulas")
    @classmethod
    def _cap_formulas(cls, v):
        from config import get_settings
        cap = get_settings().max_formulas_per_request
        if len(v) > cap:
            raise ValueError(f"自訂公式數量超過上限 ({cap})")
        return v


class TickerResult(BaseModel):
    """篩選結果中的單支股票"""
    ticker_id: str
    name: str
    market_type: Optional[str] = None
    industry: Optional[str] = None
    close: Optional[float] = None
    change_percent: Optional[float] = None
    volume: Optional[int] = None
    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    ma60: Optional[float] = None
    rsi14: Optional[float] = None
    pe_ratio: Optional[float] = None
    eps: Optional[float] = None
    foreign_buy: Optional[int] = None
    trust_buy: Optional[int] = None
    margin_balance: Optional[int] = None
    # 延伸指標
    turnover: Optional[float] = None
    avg_volume_20: Optional[float] = None
    avg_turnover_20: Optional[float] = None
    lower_shadow: Optional[float] = None
    lowest_lower_shadow_20: Optional[float] = None
    ma20_curr_month_low: Optional[float] = None
    ma20_prev_month_low: Optional[float] = None
    wma10: Optional[float] = None
    wma20: Optional[float] = None
    wma60: Optional[float] = None
    market_ok: Optional[bool] = None
    ma_bull_pullback_low_high_1_3: Optional[bool] = None
    ma_bull_pullback_low_high_2_3: Optional[bool] = None
    ma_bull_pullback_breakout_1_3: Optional[bool] = None
    ma_bull_pullback_breakout_2_3: Optional[bool] = None


class ScreenResponse(BaseModel):
    """篩選回應"""
    matched_count: int
    data: list[TickerResult]
    logic: str = "AND"
    error: Optional[str] = None
    # 篩選所依據的資料日期 (官方收盤日)。盤中可能落後即時/K線一天，供前端標示避免混淆。
    data_date: Optional[str] = None
    # 最新可用交易日 (日曆推算)。前端比對 data_date 判斷是否落後。
    latest_trading_day: Optional[str] = None
    # data_date 落後 latest_trading_day 幾個日曆天 (>1 視為過期)。
    data_age_days: Optional[int] = None
    # 資料是否過期 (data_age_days > 1)。前端據此顯示「資料非最新」提示。
    is_stale: bool = False
    # 篩選過程的非致命警告（自訂公式錯誤、規則欄位缺失/無資料等），供前端提示使用者。
    warnings: list[str] = Field(default_factory=list)


class KlineCandle(BaseModel):
    """K 線燭台資料"""
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: Optional[int] = None


class KlineResponse(BaseModel):
    """K 線資料回應"""
    ticker_id: str
    name: str
    period: str
    candles: list[KlineCandle]
    indicators: dict = Field(default_factory=dict)
